app/controllers/concerns/foreman/controller/registration.rb

```diff
@@ -82,19 +81,23 @@ def not_found(options = nil)
     false
   end
 
-  def registration_url
-    uri = if params[:url].present?
-            URI.join(params[:url], '/register')
-          else
-            URI(register_url)
-          end
+  def url
+    uri = URI(params[:url] || root_url)
 
     return uri if uri.scheme && uri.host
 
     msg = N_('URL in :url parameter is missing a scheme, please set http:// or https://')
     fail Foreman::Exception.new(msg)
   end
 
+  def context_urls
+    { url: url,
+      url_host: url.host,
+      url_port: url.port,
```

Comment:
Isn't this redundant? You already pass the `url` object. If you can't use `url.host` in the template, then passing `url` feels useless.

I do recall the last time I played with the URL object we don't have a proper jail for it. Would you mind looking into that instead?

Reply:
As you say, we don't have a jail for URL object so that's why I went this way, plus
wasn't sure if community wants to have jail for URL in the template.

I guess it could be useful, I'll add it.